Add a migration which detects and filters out unreferenced slabs

turbolent opened the pull request:
Work towards #5634

janezpodhostnik commented on cmd/util/ledger/migrations/filter_unreferenced_slabs_migration.go:
```diff
@@ -0,0 +1,138 @@
+	newPayloads := oldPayloads
+
+	err = checkStorageHealth(address, storage, oldPayloads)
+	if err != nil {
```
invert condition to reduce nesting.

turbolent replied:
Updated in 5c4feffa909b9158654bc4f754a65c7f08650c33. Left a comment why we're still returning no error at the end

turbolent commented:
@onflow/cadence PTAL 🙏

fxamacker commented on cmd/util/ledger/migrations/filter_unreferenced_slabs_migration.go:
```diff
@@ -0,0 +1,146 @@
+
+	// Filter out unreferenced root slabs.
+
+	newCount := len(oldPayloads) - len(unreferencedRootSlabIDs)
+	newPayloads = make([]*ledger.Payload, 0, newCount)
+
+	filteredPayloads := make([]*ledger.Payload, 0, len(unreferencedRootSlabIDs))
```
`unreferencedRootSlabIDs` only contains root slab IDs, so we need to filter out child slabs and referenced slabs as well.  

[PR 387](https://github.com/onflow/atree/pull/387) added a new function `(s *PersistentSlabStorage) getAllChildReferences(slab Slab)` to get all referenced slabs.  Maybe we can use it here when that function is exported.

turbolent replied:
Thanks @fxamacker! Maybe we can also integrate https://github.com/onflow/atree/pull/387 into this PR/migration in general?

fxamacker replied:
@turbolent https://github.com/onflow/atree/pull/387 is merged, so you can use `GetAllChildReferences()` after you bump atree version.

turbolent replied:
Given that this migration is for master, we'll actually need https://github.com/onflow/atree/pull/388, but no rush

fxamacker replied:
@turbolent These two PRs (for atree inlining branch) are ready for review:

- https://github.com/onflow/atree/pull/388 is updated with the same set of changes as https://github.com/onflow/atree/pull/387.  

- https://github.com/onflow/atree/pull/392 is opened (same as approved/merged PR https://github.com/onflow/atree/pull/391)

SupunS replied:
Maybe I'm missing some context here.. Is it correct to think that https://github.com/onflow/atree/pull/387 and this are two different problems? i.e: "references to non-existent registers" vs "unreferenced registers"?

turbolent replied:
@SupunS Yes, they are two separate problems. The work in e.g. https://github.com/onflow/atree/pull/387 to handle one of the problems (cleaning up missing slabs), depends on/provides a function (getting all child references) that is also needed for the other problem here (cleaning up unreferenced slabs)

fxamacker replied:
@turbolent Should we target this PR to a different branch?  `master` branch doesn't have updated atree and it might be easier if we target this to Cadence migration branch for example and I can cherry-pick it to Atree migration branch.

fxamacker replied:
@turbolent There are some breaking changes if the target is [feature/atree-inlining-cadence-v0.42](https://github.com/onflow/flow-go/tree/feature/atree-inlining-cadence-v0.42).  

Do you want to maybe target `feature/stable-cadence` first, so we don't need to change it twice due to breaking changes?  

I can port this PR to atree inlining feature branches after it is merged.

turbolent replied:
It started out targeting `feature/stable-cadence`, but then I rebased it onto `master`.

I now rebased it back, and will continue to update it (e.g. use `GetAllChildReferences`)

turbolent replied:
@fxamacker Added the call to `GetAllChildReferences` in [c2e87fd](https://github.com/onflow/flow-go/pull/5653/commits/c2e87fd0285eec48a696f9a90273bf81a5866b90). As far as I understand it is already recursing down and returns all the child slabs of the child slabs and so forth, correct?

fxamacker replied:
> Added the call to GetAllChildReferences in [c2e87fd](https://github.com/onflow/flow-go/pull/5653/commits/c2e87fd0285eec48a696f9a90273bf81a5866b90). As far as I understand it is already recursing down and returns all the child slabs of the child slabs and so forth, correct?

@turbolent Yes, also nested containers in separate slabs, etc.

fxamacker reviewed (commented):
Nice!  I left a comment about needing to filter out child slabs and referenced slabs as well.

turbolent commented:
@fxamacker This PR should be ready now, at least the original functionality, adding a migration that filters out unreferenced slabs. 

Like I mentioned above in the thread, we can maybe also extend it to take care of fixing values with missing slabs in the known affected Testnet accounts. Alternatively we could have yet another separate migration, but that would also incur some overhead. What do you think is best?

In any case, I'd prefer to get this PR an, and then take care of the additional functionality in a separate PR.
Merging this PR in it's current form would allow us to port it to the feature branch which has the v0.42 atree register inlining migration in parallel.

turbolent commented:
Forgot we also need to still integrate this new migration into the whole pipeline.

I guess it should be behind a flag, disabled by default?

fxamacker commented:
> Like I mentioned above in the thread, we can maybe also extend it to take care of fixing values with missing slabs in the known affected Testnet accounts. Alternatively we could have yet another separate migration, but that would also incur some overhead. What do you think is best?

@turbolent It would be cleaner to have separate migrations fixing different problems if overhead is small.
- fix old broken data: 10 registers with references to nonexistent registers (9 testnet accounts)
- fix old unreferenced slabs

We can make overhead small, e.g. by only fixing registers in the 9 testnet accounts.

Also, if other accounts have these problems, then it might be better to be alerted about those instead of silently fixing them.

> In any case, I'd prefer to get this PR an, and then take care of the additional functionality in a separate PR.
Merging this PR in it's current form would allow us to port it to the feature branch which has the v0.42 atree register inlining migration in parallel.

I agree.  

> Forgot we also need to still integrate this new migration into the whole pipeline.
> I guess it should be behind a flag, disabled by default?

Yes, that would be great.

turbolent commented:
Integrated the new migration into the pipeline in commit [652410e](https://github.com/onflow/flow-go/pull/5653/commits/652410e652ba743e58eb2496ed1f6261ec7bce0a).

@fxamacker To avoid the overhead of grouping payloads by account, maybe we can add the discussed migration, another account-based migration, into this sub-pipeline: https://github.com/onflow/flow-go/pull/5653/files#diff-efc0583bcfd2aa19725211a31ed4ceed33b7d0ab5228c4ab533acf0d7f474d63R422-R424

fxamacker commented:
> To avoid the overhead of grouping payloads by account, maybe we can add the discussed migration, another account-based migration, into this sub-pipeline: https://github.com/onflow/flow-go/pull/5653/files#diff-efc0583bcfd2aa19725211a31ed4ceed33b7d0ab5228c4ab533acf0d7f474d63R422-R424

@turbolent Yes, that was the plan.  I will include the new migration in an existing account based migration when I open PR.

turbolent commented:
@fxamacker Should I port this to `feature/atree-inlining-cadence-v0.42` now?

fxamacker commented:
> @fxamacker Should I port this to feature/atree-inlining-cadence-v0.42 now?

Not yet, I'm currently taking another look at this PR since it was updated.

BTW, I just opened PR #5755 for new migration (to fix refs to nonexistent slabs) targeting stable-cadence.  PTAL at #5755 :pray:

fxamacker reviewed (approved):
Nice! :+1:  LGTM!

It would be nice to write filtered slabs in payload file format (instead of JSON) to be consistent.  However, current API to create payload file doesn't support incremental writes yet.  We can revisit this aspect later if needed.

fxamacker commented:
> @fxamacker Should I port this to feature/atree-inlining-cadence-v0.42 now?

@turbolent Yes but If you don't have time, I can port this first thing in the morning (CDT).

BTW, thanks for fast review on PR #5755!

turbolent commented:
@fxamacker Can we use the reporter infrastructure to create a payloads file? Or do we have to create one payloads file per account?

fxamacker commented:
> @fxamacker Can we use the reporter infrastructure to create a payloads file? 

@turbolent No, the current reporter infrastructure can't create a payload file because it serializes received objects in JSON.  

> Or do we have to create one payloads file per account?

One payloads file per account might be overkill since we don't treat them differently.

One approach to create payload files for unreferenced slabs in this migration is to:
- aggregate unreferenced slabs in `FilterUnreferencedSlabsMigration` from all accounts
- call `CreatePayloadFile()` with all unreferenced slabs at `FilterUnreferencedSlabsMigration.Close()`

This approach uses existing code and it doesn't break any API.

Maybe report should only summarize by listing accounts having any unreferenced slabs (along with undreferenced slab count).  Payload data can be in a payloads file instead of the report.

fxamacker commented:
@turbolent I opened PR #5758 to port this PR to `feature/atree-inlining-cadence-v0.42`.  PTAL :pray:

turbolent commented:
@fxamacker How does [7e3d598](https://github.com/onflow/flow-go/pull/5653/commits/7e3d5981ab24cf31d1a367840fa36c42d30be037) look?

turbolent commented on cmd/util/ledger/migrations/filter_unreferenced_slabs_migration.go:
```diff
@@ -136,25 +147,76 @@ func (m *FilterUnreferencedSlabsMigration) MigrateAccount(
+		m.filteredPayloads,
+		m.filteredAccounts,
+		true,
+	)
```
Oh, I didn't realize that the parameter was used to filter the payloads, I assumed it was just metadata.

Removed in f5a46ee35b4fb609642dc83bb071e7ffbc36578b

turbolent commented on cmd/util/ledger/migrations/filter_unreferenced_slabs_migration.go:
```diff
@@ -25,19 +29,26 @@ func StorageIDFromRegisterID(registerID flow.RegisterID) atree.StorageID {
+	rw               reporters.ReportWriter
+	outputDir        string
+	mutex            sync.Mutex
+	filteredAccounts []common.Address
```
Removed in f5a46ee35b4fb609642dc83bb071e7ffbc36578b

turbolent commented on cmd/util/ledger/migrations/filter_unreferenced_slabs_migration.go:
```diff
@@ -25,19 +29,26 @@ func StorageIDFromRegisterID(registerID flow.RegisterID) atree.StorageID {
+	outputDir        string
+	mutex            sync.Mutex
+	filteredAccounts []common.Address
+	filteredPayloads []*ledger.Payload
```
Done in 52175eab577066169d9f24560b22a5a907701869

turbolent commented:
@fxamacker Addressed the feedback in f5a46ee35b4fb609642dc83bb071e7ffbc36578b and 52175eab577066169d9f24560b22a5a907701869. If they look good, we'll want to port them to #5758

fxamacker commented:
@turbolent Updates look good!  I ported new commits to #5758.
